### ISSUE TITLE
Does spspmm operation support autograd?

### ISSUE THREAD

**Issue by changym3** [rusty1s/pytorch_sparse#45]
Hi,  you say only autograd support is supported for values tensors, but it seems it doesn't work in spspmm.

Like this:
`indexA = torch.tensor([[0, 0, 1, 2, 2], [1, 2, 0, 0, 1]])
valueA = torch.tensor([1, 2.0, 3, 4, 5], requires_grad=True)

indexB = torch.tensor([[0, 2], [1, 0]])
valueB = torch.tensor([2, 4.0], requires_grad=True)

indexC, valueC = torch_sparse.spspmm(indexA, valueA, indexB, valueB, 3, 3, 2)

print(valueC.requires_grad)
print(valueC.grad_fn)`

In my case, I want to parameterize the sparse adjacent matrix and feature matrix in GCN, so the inputs need to be both differentiable. I wonder if there're some bugs or just the way it is.

Regards.

**Comment by rusty1s** [comment]
That's the only function that does not have proper autograd support. Gradients for sparse-sparse matrix multiplication are quite difficult to obtain (since they are usually dense). I had a working, but slow implementation up to `0.4.4` release, but removed it since it wasn't a really good implementation. If you desperately need it, feel free to try it out.

**Comment by changym3** [comment]
> That's the only function that does not have proper autograd support. Gradients for sparse-sparse matrix multiplication are quite difficult to obtain (since they are usually dense). I had a working, but slow implementation up to `0.4.4` release, but removed it since it wasn't a really good implementation. If you desperately need it, feel free to try it out.

Hey! Thanks for your great work! I have installed the `0.4.4` release of `torch_sparse` and it totally works out in my experiments! Maybe you could add this information to the documentation. It takes me so long to figure out this no-autograd problem.

Thanks a lot again!

**Comment by LuciusMos** [comment]
> Hey! Thanks for your great work! I have installed the `0.4.4` release of `torch_sparse` and it totally works out in my experiments! Maybe you could add this information to the documentation. It takes me so long to figure out this no-autograd problem.
> 
> Thanks a lot again!

Thank you so much for your question raising! It really troubles me for almost a week!

**Comment by rusty1s** [comment]
Sorry for the inconveniences. I have plans to add `backward` support for `spspmm` back in, see https://github.com/rusty1s/pytorch_geometric/issues/1465.

**Comment by jlevy44** [comment]
Do you have any updates on autograd support?

**Comment by jlevy44** [comment]
I'm parameterizing the weights of a sparse matrix to treat it as a locally connected network for a sparsely connected MLP implementation. Could I still run a backward pass to update these weights after calling matmul between this sparse matrix and a dense input?

**Comment by jlevy44** [comment]
Nevermind, already seeing some nice implementations out there!
https://pypi.org/project/sparselinear/
https://stackoverflow.com/questions/63893602/neural-network-layer-without-all-connections

**Comment by JRD971000** [comment]
Does spspmm still lack autograd support?

**Comment by jaynanavati-az** [comment]
Does spspmm still lack autograd support?

**Comment by rusty1s** [comment]
Sadly yes :(

**Comment by rusty1s** [comment]
There isn‘t a workaround except for installing an earlier version. If you are interested, we can try to bring it back with your help. WDYT?

**Comment by jaynanavati-az** [comment]
@rusty1s sounds good, why don't we start with putting back your existing implementation? is it not better than having nothing?

**Comment by rusty1s** [comment]
Here's the roadmap in order to achieve this:
* [ ] Add the CPU `spspmm_bw` code from https://github.com/rusty1s/pytorch_sparse/blob/0.4.4/cpu/spspmm.cpp#L19-L63
* [ ] Add the CUDA `spspmm_bw` code from https://github.com/rusty1s/pytorch_sparse/blob/0.4.4/cuda/spspmm_kernel.cu#L115-L166
* [ ] Register the `autograd` function as in https://github.com/rusty1s/pytorch_sparse/blob/master/csrc/spmm.cpp#L55-L86
* [ ] Add a basic `backward` test as in https://github.com/rusty1s/pytorch_sparse/blob/0.4.4/test/test_spspmm.py#L29-L35

**Comment by jaynanavati-az** [comment]
Is there an alternative? It is difficult to get earlier versions of torch sparse that have this to work on newer cuda versions.. :(
________________________________

**Comment by rusty1s** [comment]
With PyTorch 1.12, I assume you can also try to use the sparse-matrix multiplication from PyTorch directly. PyTorch recently integrated better sparse matrix support into its library :)
